Remove guard_size_oblivious from vector_norm decomposition.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148809

This PR remove the usage of guard_size_oblivious in vector_norm by inlining it in the runtime check, 
this prevent any data dependent error from ever appearing here at the locations where guard_size_oblivious
used to exist. Before this PR it used to break potentially. This is NOT BC breaking or changing of semantics from eager. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

laithsakka committed:
remove gaurd_size_oblivious from vector_norm

[ghstack-poisoned]

laithsakka committed:
Update on "remove gaurd_size_oblivious from vector_norm"




[ghstack-poisoned]

laithsakka committed:
Update on "[DRAFT] remove guard_size_oblivious from vector_norm"




[ghstack-poisoned]

laithsakka committed:
Update on "[DRAFT] remove guard_size_oblivious from vector_norm"




[ghstack-poisoned]

laithsakka committed:
Update on "[DRAFT] remove guard_size_oblivious from vector_norm"




[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm"



[ghstack-poisoned]

bobrenjc93 commented:
Can you add a test? Especially for these first few migrations it’d be good to ensure it’s doing what we expect.

aorenste commented on torch/_refs/linalg/__init__.py:
```diff
@@ -107,29 +108,36 @@ def vector_norm(
+            shape = x.shape
+            assert (
+                dim is not None
+            )  # mypy does not seem to be able to see through check?
```
It's true that mypy doesn't understand torch._check() - but also your check isn't guaranteeing that dim is not None, is it? (If `x.numel() != 0` then `dim` could be `None`)

laithsakka replied:
this is exiting code i did not add this

laithsakka replied:
but anyway i am changing this will add an if

aorenste commented on torch/_refs/linalg/__init__.py:
```diff
@@ -107,29 +108,36 @@ def vector_norm(
+                "because the operation does not have an identity",
             )
+            shape = x.shape
+            assert (
+                dim is not None
+            )  # mypy does not seem to be able to see through check?
```
I'm sure pyfmt reflowed this onto three lines...
```suggestion
            # mypy does not seem to be able to see through check?
            assert dim is not None
```

laithsakka replied:
i am removing this assertion will add if instead.

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented on torch/_refs/linalg/__init__.py:
```diff
@@ -61,6 +61,7 @@ def _check_norm_dtype(dtype: Optional[torch.dtype], x_dtype: torch.dtype, fn_nam
 
 
 import operator
+from operator import or_ as sym_or
```
i just want to call operator.or_ but want it named sym_or cause its what would happen.

laithsakka replied:
i can add sym_or = operator._or to the symbolic shapes and add a comment about it for future clairty

laithsakka commented on torch/_refs/linalg/__init__.py:
```diff
@@ -107,29 +108,36 @@ def vector_norm(
-                lambda: f"linalg.vector_norm cannot compute the {ord} norm on the "
-                f"dimension {d} because this dimension is empty and the "
-                "operation does not have an identity",
+                sym_or(x.numel() != 0, dim is not None and len(dim) != 0),
```
will force guarding.

laithsakka replied:
you can try this example
```
class Mod(torch.nn.Module):
    def forward(self,a: torch.Tensor, b: torch.Tensor):
        x = a.item()
        torch._check(x>10 or True)

        # torch._check(operator.or_( x>10 , True))
        return b*10
    

a = torch.tensor([1])
b = torch.torch.ones(10, 10, device="cuda",dtype=torch.float64) 

exported_program: torch.export.ExportedProgram = torch.export.export(
    Mod(), args=(a,b))
print(exported_program.graph_module.code.strip())
```

eellison commented on test/test_linalg.py:
```diff
@@ -1445,6 +1447,46 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
+                m , args= tuple(torch.tensor([1])))
+            print(exported_program.graph_module.code.strip())     
+
+        # # Those are situation where we should throw error if input.size()==0
```
Why do we have these commented out ? if its for something that should work, but doesnt, you can add a test with expected failure.

laithsakka replied:
ah this is draft i am adding unit tests for got to mark the PR draft

eellison reviewed (commented):
Some of the decomps are for numerics, which I think we should guard for. see http://github.com/pytorch/pytorch/pull/144073/

cc @isuruf

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1277,6 +1277,12 @@ def statically_known_true(x: Union[bool, SymBool]) -> bool:
+# a or b is eager, meaning that (a==1 or True) will specialize a.
+# but operator.or_(a==1, True) generates sympy expression and avoid
+# the specialization.
+sym_or = operator.or_
```
this already exists on SymNode https://www.internalfb.com/code/fbsource/[6f1893cbc5666e46b21f179a39bea4d5df2fd462]/xplat/caffe2/torch/fx/experimental/sym_node.py?lines=439

bobrenjc93 commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1277,6 +1277,12 @@ def statically_known_true(x: Union[bool, SymBool]) -> bool:
 
+# a or b is eager, meaning that (a==1 or True) will specialize a.
+# but operator.or_(a==1, True) generates sympy expression and avoid
+# the specialization.
```
It took me some time to understand this. The key is that operator.or_ will respect the dunder method `__or__` on symnode instead of eagerly evaluating. Perhaps you can update the comment to reflect this?

aorenste replied:
Agreed - Definitely need to explain the difference between `or` and `operator.or_` for symnodes...

laithsakka replied:
i will add.

bobrenjc93 commented on test/test_linalg.py:
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
+        # dim is None
+        test(-1, None, True)
+
+        # len(dim) ==0
```
s/==0/== 0

bobrenjc93 commented on test/test_linalg.py:
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
+        # shape[d] == 0
+        test(-1, [0], False, True)
+
+        # u0+1==0 is False so we do not see a runtime assert
```
s/u0+1==0/u0 + 1 == 0

bobrenjc93 replied:
s/so we do not see a runtime assert/so we do not expect to fail on the runtime assert

laithsakka replied:
mm i meant we do not see runtime assert in the generated graph which was i this testing

bobrenjc93 commented:
Overall looks good. One thing that'd be helpful for reviewers on these migration PRs is to include summaries with a sense of the spirit and risk of the change. Are we introducing any BC changes? What new checks did we add and why? Do we expect more data-dependent errors or less? etc. It might also make sense to add comments in the meta functions themselves so future readers have an easier time understanding the intention behind these guards/checks.

cc @pianpwk

aorenste commented on torch/_refs/linalg/__init__.py:
```diff
@@ -97,6 +97,34 @@ def diagonal(
+            x.numel() != 0,
+            not isinstance(dim, IntLike) and dim is not None and len(dim) != 0,
+        ),
+        "linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
```
```suggestion
        lambda: (f"linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
```

laithsakka replied:
this is intentional torch.compile does not support lambda strings here.
https://github.com/pytorch/pytorch/issues/149997

aorenste commented on torch/_refs/linalg/__init__.py:
```diff
@@ -97,6 +97,34 @@ def diagonal(
+        for d in dim:
+            torch._check(
+                sym_or(x.numel() != 0, d < len(shape) and d >= 0 and shape[d] != 0),
+                "linalg.vector_norm cannot compute the {ord} norm on the "
+                f"dimension {d} because this dimension is empty and the "
+                "operation does not have an identity",
```
You don't want to use an f-string outside of a lambda (because it will compute the string even if the check passes)

```suggestion
                lambda: (f"linalg.vector_norm cannot compute the {ord} norm on the "
                f"dimension {d} because this dimension is empty and the "
                "operation does not have an identity"),
```

laithsakka replied:
this is intentional torch.compile does not support lambda strings here. 
https://github.com/pytorch/pytorch/issues/149997

laithsakka replied:
what we can do , is have a utility that pre check to see if the compile check will be deferred or not, and then calls this or that, or fix the issue above, buts its not related to this change. 
previoulsy it was ok, because we would never defer those runtime asserts. not since we can defer them we hit this limitation.

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."


This PR remove the usage of guard_size_oblivious in vector_norm by inlining it in the runtime check, 
this prevent any data dependent error from ever appearing here at the locations where guard_size_oblivious
used to exist. Before this PR it used to break potentially. This is NOT BC breaking or changing of semantics from eager. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."


This PR remove the usage of guard_size_oblivious in vector_norm by inlining it in the runtime check, 
this prevent any data dependent error from ever appearing here at the locations where guard_size_oblivious
used to exist. Before this PR it used to break potentially. This is NOT BC breaking or changing of semantics from eager. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented on test/test_linalg.py:
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
+                expect_index_0_check,
+            )
+
+        # dim is int.
```
remove the `.` or add punctuation to the rest?

bobrenjc93 commented on test/test_linalg.py:
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
+        # shape[d] == 0
+        test(-1, [0], False, True)
+
+        # u0+1 == 0 is False we do not see a runtime assert in the generated graph.
```
`u0 + 1`, notice the spaces

bobrenjc93 commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1277,6 +1277,19 @@ def statically_known_true(x: Union[bool, SymBool]) -> bool:
+# When the whole expression is passed to evaluation in that case, we do not throw a
+# data dependent error or guard because we can statically know the result is True
+# before unpacking the symbols.
+sym_or = operator.or_
```
given that @pian is introducing sym_or in https://github.com/pytorch/pytorch/pull/150456#event-17082477541, does it make sense to rebase on top of his PR?

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on " Remove guard_size_oblivious from vector_norm decomposition."


This PR remove the usage of guard_size_oblivious in vector_norm by inlining it in the runtime check, 
this prevent any data dependent error from ever appearing here at the locations where guard_size_oblivious
used to exist. Before this PR it used to break potentially. This is NOT BC breaking or changing of semantics from eager. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot merge -i
